### PR TITLE
Port get-file-char

### LINKED ISSUE #1300: Port get-file-char
```c
DEFUN ("get-file-char", Fget_file_char, Sget_file_char, 0, 0, 0,
       doc: /* Don't use this yourself.  */)
  (void)
{
  if (!infile)
    error ("get-file-char misused");
  return make_number (readbyte_from_stdio ());
}
```

### PR DESCRIPTION
**Pull request by agraven**
I was on the fence on whether or not to port `readbyte_from_stdio`, as it relies on a lot of behaviour that's expressed much more elegantly in c than rust, but I think i managed to get it right.

Closes #1300.

### CODE REVIEW

File: rust_src/src/lread.rs [hunk 1]
```diff
@@ -155,6 +159,41 @@ pub unsafe fn defvar_per_buffer_offset(
...
+        let idx = (*infile.sub(1)).lookahead as usize;
+        return (*infile).buf[idx].into();
+    }
+
+    let instream = (*infile).stream;
+
+    block_input();
+
+    // Interrupted read have been observed while reading over the network.
+    let mut c;
```
Comment (shaleh) [inline rust_src/src/lread.rs]: An alternative implementation that might be more idiomatic Rust is:

    loop {
        c = getc_unlocked(instream);
        if c == libc::EOF && errno().0 == libc::EINTR && ferror_unlocked(instream) > 0 {
            unblock_input();
            maybe_quit();
            block_input();
            clearerr_unlocked(instream);
        } else {
            break;
        }
    }

If we factor out the check....

    fn is_interrupted_read(stream: TheStreamType, value: i32) -> bool {
        c == libc::EOF && errno().0 == libc::EINTR && ferror_unlocked(stream) > 0
    }

    loop {
        c = getc_unlocked(instream);
        if not is_interrupted_read(instream, c) {
            break;
        }
        unblock_input();
        maybe_quit();
        block_input();
        clearerr_unlocked(instream);
    }

hmmm

    fn interruptable_read(stream: TheStreamType) -> i32 {
        loop {
            let c = getc_unlocked(stream);
            if c == libc::EOF && errno().0 == libc::EINTR && ferror_unlocked(instream) > 0 {
                unblock_input();
               maybe_quit();
               block_input();
               clearerr_unlocked(instream);
           } else {
                return c;
           }
        }
    }

    let c = interruptable_read(instream);
    
    unblock_input();

    if c != libc::EOF {
        c
    } else {
        -1
    }

File: rust_src/src/lread.rs [hunk 2]
```diff
@@ -155,6 +159,41 @@ pub unsafe fn defvar_per_buffer_offset(
     }
 }
 
+/// Read a byte from stdio. If it has lookahead, use the stored value.
+/// If read over network is interrupted, keep trying until read succeeds.
+#[no_mangle]
+pub unsafe extern "C" fn readbyte_from_stdio() -> i32 {
+    // If infile has lookahead, use stored value
+    if (*infile).lookahead > 0 {
+        let idx = (*infile.sub(1)).lookahead as usize;
```
Comment (shaleh) [inline rust_src/src/lread.rs]: This does not have the same behavior. The C code is decrementing `lookahead` not `infile`. It should read:

    if (*infile).lookahead > 0 {
        (*infile).lookahead -= 1;
        return (*infile).buf[(*infile).lookahead as usize];
    }

Comment (agraven) [reply rust_src/src/lread.rs]: That makes way more sense, I thought the way I read it seemed off.

Comment (agraven) [reply rust_src/src/lread.rs]: Not having to do pointer math also means I can convert the pointer to a reference, which means I won't have to dereference it explicitly so many times.

File: rust_src/src/lread.rs (hunk 1 above)
Comment (agraven) [reply rust_src/src/lread.rs]: The loop is definitely a more readable approach, but I'm inclined to leave the code in the function body as it's not used anywhere else, and I consider the slight readability trade-off worth having the code in one place.

Comment (shaleh) [reply rust_src/src/lread.rs]: Fair enough.

File: rust_src/src/lread.rs [hunk 3]
```diff
@@ -155,6 +159,44 @@ pub unsafe fn defvar_per_buffer_offset(
...
+    if file.lookahead > 0 {
+        file.lookahead -= 1;
+        return file.buf[file.lookahead as usize].into();
+    }
+
+    let instream = file.stream;
+
+    block_input();
+
+    // Interrupted read have been observed while reading over the network.
```
Comment (shaleh) [inline rust_src/src/lread.rs]: Nice. I was going to suggest this format but wrote out the function/return version instead.

File: rust_src/src/lread.rs [hunk 4]
```diff
@@ -155,6 +159,44 @@ pub unsafe fn defvar_per_buffer_offset(
     }
 }
 
+/// Read a byte from stdio. If it has lookahead, use the stored value.
+/// If read over network is interrupted, keep trying until read succeeds.
+#[no_mangle]
+pub unsafe extern "C" fn readbyte_from_stdio() -> i32 {
```
Comment (shaleh) [inline rust_src/src/lread.rs]: Definite improvement in readability. :+1:

File: rust_src/src/lread.rs [hunk 5]
```diff
@@ -192,6 +234,15 @@ pub fn read(stream: LispObject) -> LispObject {
     }
 }
 
+/// Don't use this yourself
```
Comment (shaleh) [inline rust_src/src/lread.rs]:
```suggestion
/// Don't use this yourself.
```

The end punctuation was dropped.

**Pull request merged**